In [123]:
%pylab inline
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import csv

# Importing and reading CSV file
file_input = 'C:\\Users\\Sneha Rani\\CapstoneProject-LR\\ibm-hr-attrition.csv'
df1 = pd.read_csv(file_input, index_col=None)

# Pre cleaned the dataset by selecting significant columns
df2 = df1[['Age','Attrition','BusinessTravel','Department','DistanceFromHome','Education','EducationField',
           'EnvironmentSatisfaction','Gender','HourlyRate','JobInvolvement','JobRole','JobSatisfaction','MaritalStatus','MonthlyIncome',
           'NumCompaniesWorked','OverTime','PercentSalaryHike','RelationshipSatisfaction','TotalWorkingYears',
           'WorkLifeBalance','YearsAtCompany','YearsInCurrentRole','YearsSinceLastPromotion','YearsWithCurrManager']]

# Converting strings to categorical values
from sklearn.preprocessing import LabelEncoder
number = LabelEncoder()
df2['Attrition'] = number.fit_transform(df2['Attrition'].astype('str'))
df2.head()

Populating the interactive namespace from numpy and matplotlib


F:\Anaconda3\lib\site-packages\IPython\core\magics\pylab.py:161: UserWarning: pylab import has clobbered these variables: ['number']
`%matplotlib` prevents importing * from pylab and numpy
  "\n`%matplotlib` prevents importing * from pylab and numpy"
F:\Anaconda3\lib\site-packages\ipykernel\__main__.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


Age  Attrition     BusinessTravel              Department  \
0   41          1      Travel_Rarely                   Sales   
1   49          0  Travel_Frequently  Research & Development   
2   37          1      Travel_Rarely  Research & Development   
3   33          0  Travel_Frequently  Research & Development   
4   27          0      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EnvironmentSatisfaction  \
0                 1          2  Life Sciences                        2   
1                 8          1  Life Sciences                        3   
2                 2          2          Other                        4   
3                 3          4  Life Sciences                        4   
4                 2          1        Medical                        1   

   Gender  HourlyRate          ...           NumCompaniesWorked OverTime  \
0  Female          94          ...                            8      Yes   
1    Male          61          ...                            1       No   
2    Male          92          ...                            6      Yes   
3  Female          56          ...                            1      Yes   
4    Male          40          ...                            9       No   

   PercentSalaryHike RelationshipSatisfaction  TotalWorkingYears  \
0                 11                        1                  8   
1                 23                        4                 10   
2                 15                        2                  7   
3                 11                        3                  8   
4                 12                        4                  6   

   WorkLifeBalance YearsAtCompany  YearsInCurrentRole  \
0                1              6                   4   
1                3             10                   7   
2                3              0                   0   
3                3              8                   7   
4                3              2                   2   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  
3                        3                     0  
4                        2                     2  

[5 rows x 25 columns]

In [124]:
# Replacing numeric categorical features data with categorical values

grade = {"Education":{1: "Below College", 2:"College", 3:"Bachelor", 4:"Master", 5:"Doctor"},
       "EnvironmentSatisfaction":{1: "Low", 2:"Medium", 3:"High", 4:"Very High"},
       "JobInvolvement":{1:"Low", 2:"Medium", 3:"High", 4:"Very High"}, 
       "JobSatisfaction":{1:"Low", 2:"Medium", 3:"High", 4:"Very High"},  
       "RelationshipSatisfaction":{1:"Low", 2:"Medium", 3:"High", 4:"Very High"},
       "WorkLifeBalance":  {1: "Bad", 2:"Good", 3:"Better", 4:"Best"},
      }
df2.replace(grade, inplace=True)
df2.head()

F:\Anaconda3\lib\site-packages\pandas\core\generic.py:3855: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  regex=regex)


Age  Attrition     BusinessTravel              Department  \
0   41          1      Travel_Rarely                   Sales   
1   49          0  Travel_Frequently  Research & Development   
2   37          1      Travel_Rarely  Research & Development   
3   33          0  Travel_Frequently  Research & Development   
4   27          0      Travel_Rarely  Research & Development   

   DistanceFromHome      Education EducationField EnvironmentSatisfaction  \
0                 1        College  Life Sciences                  Medium   
1                 8  Below College  Life Sciences                    High   
2                 2        College          Other               Very High   
3                 3         Master  Life Sciences               Very High   
4                 2  Below College        Medical                     Low   

   Gender  HourlyRate          ...          NumCompaniesWorked OverTime  \
0  Female          94          ...                           8      Yes   
1    Male          61          ...                           1       No   
2    Male          92          ...                           6      Yes   
3  Female          56          ...                           1      Yes   
4    Male          40          ...                           9       No   

  PercentSalaryHike RelationshipSatisfaction  TotalWorkingYears  \
0                11                      Low                  8   
1                23                Very High                 10   
2                15                   Medium                  7   
3                11                     High                  8   
4                12                Very High                  6   

   WorkLifeBalance YearsAtCompany  YearsInCurrentRole YearsSinceLastPromotion  \
0              Bad              6                   4                       0   
1           Better             10                   7                       1   
2           Better              0                   0                       0   
3           Better              8                   7                       3   
4           Better              2                   2                       2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 25 columns]

In [125]:
# Creating dummy variables on categorical data for easy interpretation

final_data = pd.get_dummies(df2, columns =['BusinessTravel', 'Department', 'Education', 'EducationField', 'EnvironmentSatisfaction',
                                     'Gender', 'JobInvolvement', 'JobRole', 'JobSatisfaction', 'MaritalStatus', 'OverTime', 
                                     'RelationshipSatisfaction', 'WorkLifeBalance'])
final_data.head()

Age  Attrition  DistanceFromHome  HourlyRate  MonthlyIncome  \
0   41          1                 1          94           5993   
1   49          0                 8          61           5130   
2   37          1                 2          92           2090   
3   33          0                 3          56           2909   
4   27          0                 2          40           3468   

   NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  YearsAtCompany  \
0                   8                 11                  8               6   
1                   1                 23                 10              10   
2                   6                 15                  7               0   
3                   1                 11                  8               8   
4                   9                 12                  6               2   

   YearsInCurrentRole          ...           OverTime_No  OverTime_Yes  \
0                   4          ...                     0             1   
1                   7          ...                     1             0   
2                   0          ...                     0             1   
3                   7          ...                     0             1   
4                   2          ...                     1             0   

   RelationshipSatisfaction_High  RelationshipSatisfaction_Low  \
0                              0                             1   
1                              0                             0   
2                              0                             0   
3                              1                             0   
4                              0                             0   

   RelationshipSatisfaction_Medium  RelationshipSatisfaction_Very High  \
0                                0                                   0   
1                                0                                   1   
2                                1                                   0   
3                                0                                   0   
4                                0                                   1   

   WorkLifeBalance_Bad  WorkLifeBalance_Best  WorkLifeBalance_Better  \
0                    1                     0                       0   
1                    0                     0                       1   
2                    0                     0                       1   
3                    0                     0                       1   
4                    0                     0                       1   

   WorkLifeBalance_Good  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 65 columns]

In [126]:
# Move the reponse variable "Attrition" to the front of the table

front = final_data['Attrition']
final_data.drop(labels=['Attrition'], axis=1, inplace = True)
final_data.insert(0, 'Attrition', front)
final_data.head()

Attrition  Age  DistanceFromHome  HourlyRate  MonthlyIncome  \
0          1   41                 1          94           5993   
1          0   49                 8          61           5130   
2          1   37                 2          92           2090   
3          0   33                 3          56           2909   
4          0   27                 2          40           3468   

   NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  YearsAtCompany  \
0                   8                 11                  8               6   
1                   1                 23                 10              10   
2                   6                 15                  7               0   
3                   1                 11                  8               8   
4                   9                 12                  6               2   

   YearsInCurrentRole          ...           OverTime_No  OverTime_Yes  \
0                   4          ...                     0             1   
1                   7          ...                     1             0   
2                   0          ...                     0             1   
3                   7          ...                     0             1   
4                   2          ...                     1             0   

   RelationshipSatisfaction_High  RelationshipSatisfaction_Low  \
0                              0                             1   
1                              0                             0   
2                              0                             0   
3                              1                             0   
4                              0                             0   

   RelationshipSatisfaction_Medium  RelationshipSatisfaction_Very High  \
0                                0                                   0   
1                                0                                   1   
2                                1                                   0   
3                                0                                   0   
4                                0                                   1   

   WorkLifeBalance_Bad  WorkLifeBalance_Best  WorkLifeBalance_Better  \
0                    1                     0                       0   
1                    0                     0                       1   
2                    0                     0                       1   
3                    0                     0                       1   
4                    0                     0                       1   

   WorkLifeBalance_Good  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  

[5 rows x 65 columns]

In [127]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import scale

# Defining the first parameter 

X = final_data.iloc[:, 1:64].values
print("X: ", type(X), X.shape)

X:  <class 'numpy.ndarray'> (1470, 63)


In [128]:
# Defining the second parameter 

y = final_data.iloc[:, 0].values
print("y: ", type(y), y.shape)

y:  <class 'numpy.ndarray'> (1470,)


### Logistic Regression Base Line Model using 'L2' regularization

In [140]:
# Split the data into training and test sets
# Stratify parameter makes a split so that proportion of values in the sample produced will be same as proportion of values
# provided to parameter stratify.
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=50, stratify=y)

# Build the Logistic Regression Model using default 'L2' regularization

classifier = LogisticRegression()

# Fit the model on the trainng data.
classifier.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [145]:
# Print the accuracy from the testing data.
y_predict_test = classifier.predict(X_test)
print("\n")
print("[Test] Accuracy score:" ,accuracy_score(y_predict_test, y_test))

# Print the accuracy from the training data.
y_predict_train = classifier.predict(X_train)
print("\n")
print("[Training] Accuracy score:" ,accuracy_score(y_predict_train, y_train))



[Test] Accuracy score: 0.872282608696


[Training] Accuracy score: 0.896551724138


From the above results, I see that the **Test Accuracy score: 0.872282608696** is pretty close to 
**Training Accuracy score: 0.896551724138** using default **L2** regularization which is good. 

As one says, that when the training accuracy is good, and the test accuracy is not close to the training accuracy (in the sense that test accuracy is worse), then the model is "overfitting" the training set.

So, there is no overfitting in this case.
Also, this shows that model will generalize well when presented with new sets of data as training accuracy score is slightly greater than test accuracy score which means there is less chance of getting  a bug.

**I will dive more deeply and evaluate the performance of binary classifiers by computing a confusion matrix and generating a classification report.**

In [146]:
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html
# confusion matrix for test data using default 'L2' regularization
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_predict_test)
confusion_matrix

array([[298,  11],
       [ 36,  23]])

*Confusion matrix for test data using 'L2' regularization tells us that 298+23 are correct predictions and 36+11 are incorrect predictions.*

In [139]:
# Classification report using 'L2' regularization 

from sklearn.metrics import classification_report
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report

print("[Training Classification Report:]")
print(classification_report(y_train, y_predict_train))

print("[Test Classification Report:]")
print(classification_report(y_test, y_predict_test))

[Training Classification Report:]
             precision    recall  f1-score   support

          0       0.90      0.98      0.94       924
          1       0.83      0.46      0.59       178

avg / total       0.89      0.90      0.88      1102

[Test Classification Report:]
             precision    recall  f1-score   support

          0       0.89      0.96      0.93       309
          1       0.68      0.39      0.49        59

avg / total       0.86      0.87      0.86       368



From above Classification Report, we find that training performance metrics are--as expected--slightly better than their test set counterpart.

### Logistic Regression Base Line Model using 'L1' regularization

In [147]:
# Build the Logistic Regression Model using 'L1' regularization
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html

classifier_1 = LogisticRegression(penalty = 'l1')

# Fit the model on the trainng data.
classifier_1.fit(X_train, y_train)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l1', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [149]:
# Checking the model's accuracy from the testing data.
y_predict_test_1 = classifier_1.predict(X_test)
print("\n")
print("[Test] Accuracy score:" ,accuracy_score(y_predict_test_1, y_test))

# Checking the model's accuracy from the training data.
y_predict_train_1 = classifier_1.predict(X_train)
print("\n")
print("[Training] Accuracy score:" ,accuracy_score(y_predict_train_1, y_train))



[Test] Accuracy score: 0.872282608696


[Training] Accuracy score: 0.893829401089


From the above results, I see that the **Test Accuracy score: 0.872282608696** is pretty close to 
**Training Accuracy score: 0.893829401089** using **L1** regularization which is quite good in this case also.

So, there is no overfitting in this case.
Also, this shows that model will generalize well when presented with new sets of data as training accuracy score is slightly greater than test accuracy score which means there is less chance of getting  a bug.

In [150]:
# confusion matrix for test data using 'L1' regularization
from sklearn.metrics import confusion_matrix
confusion_matrix = confusion_matrix(y_test, y_predict_test)
confusion_matrix

array([[298,  11],
       [ 36,  23]])

*Confusion matrix for test data using 'L1' regularization tells us that 298+23 are correct predictions and 36+11 are incorrect predictions.*

In [151]:
# Classification report using 'L1' regularization 

from sklearn.metrics import classification_report
# ref: http://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report

print("[Training Classification Report:]")
print(classification_report(y_train, y_predict_train_1))

print("[Test Classification Report:]")
print(classification_report(y_test, y_predict_test_1))

[Training Classification Report:]
             precision    recall  f1-score   support

          0       0.90      0.98      0.94       924
          1       0.82      0.44      0.57       178

avg / total       0.89      0.89      0.88      1102

[Test Classification Report:]
             precision    recall  f1-score   support

          0       0.89      0.96      0.93       309
          1       0.68      0.39      0.49        59

avg / total       0.86      0.87      0.86       368



*From above Classification Report, we find that training performance metrics are--as expected--slightly better than their test set counterpart.*